### PR TITLE
Fix #41413 by using a specific purpose for direct uploads

### LINKED ISSUE #41413: ActiveStorage blobs can be attached using their "displayable" signed IDs
### Steps to reproduce

ActiveStorage blobs currently use signed IDs. The purpose however is always the same (blob_id). When uploading a Blob through the direct_upload endpoint, a user gets the same signed ID for a Blob as the one that is later used for attaching a Blob to a model through Attached - see https://github.com/rails/rails/blob/bddb2c9b193c976dbe6a9a1358479d1792931986/activestorage/app/controllers/active_storage/direct_uploads_controller.rb#L18

Since the signed blob ID is the same as the one used for later retrieving a blob (even for a small period of time, since the purpose of the signature is the same), the following attack can be performed:

* Alice uploads a PDF document and shares a link to a page with it with Bob
* Mallory manages to view that page for whatever reason
* Mallory examines the source page and copies the signed ID of the blob on that page
* Mallory then uses her account on the same service and supplies the signed ID of the blob in a form POST request which would perform file attachment and create a reference to the blob in their account too.

At that stage, the Blob in the database would have more referencing Attachments since an Attachment would be generated allowing Mallory to have full access to the Blob in their account on the service. This would also prevent the Blob from being garbage-collected if Alice decides to remove it from "her" attachment. At first glance, I see the following possible impact:

* An attacker can link attachments stored on the service into the model objects belonging to the attacker, even though they weren't uploaded by the attacker
* An attacker can "increase refcount" for Blobs and thus prevent those blobs from being deleted from the application once there are no Attachments referencing them from records belonging to accounts that did have access.

Mitigation would possibly require using a different signature purpose for direct uploading / attaching than the rest of the ActiveStorage operations. An alternative would be to blend the user account identifier into the signature for upload, so that an account may only use signed IDs which were generated for them / on their behalf (Since Rails does not provide authorization out-of-the-box I don't see that likely but might be possible).

### Expected behavior

It is only possible to attach a blob to a record using a signed ID if that signed ID was issued to you for the purpose of upload (proof-of-work).

### Actual behavior

Signed IDs are used interchangeably and attaching by any user is possible, even if the user was not the original uploader.

### System configuration
**Rails version**: 5.x -6.x

**Ruby version**: any relevant version for framework versions mentioned above

### PR DESCRIPTION
**Pull request by santib**
### Summary

Fixes #41413 by using a specific `purpose` when generating the `signed_id` of `ActiveStorage::Blob`s for direct uploads. I think it would be important to fix it since a user should have the right to permanently delete blobs uploaded by them.

Before this change, the attack described in the issue was possible just by knowing the `signed_id` of a Blob. And the `signed_id` is present in shareable URLs, so it's expected that some people know some `signed_id`s.

After this change, the attack described would only be possible if intercepting the response of the direct uploads controller. Which wouldn't be possible just because of SSL.

This is a **breaking change** since it'd break any app that performs any `find_signed` call in the context of direct uploads because it's assuming the purpose would be `blob_id`.

cc @georgeclaghorn

### CODE REVIEW

File: activestorage/app/models/active_storage/blob.rb [hunk 1]
```diff
@@ -155,6 +155,11 @@ def signed_id(purpose: :blob_id)
     super
   end
 
+  # Returns a signed ID for this blob that's suitable for direct uploads.
+  def direct_upload_signed_id(purpose: :direct_upload)
+    signed_id(purpose: purpose)
+  end
+
```
Comment (georgeclaghorn) [inline activestorage/app/models/active_storage/blob.rb:162]: Let’s get rid of this method and use `signed_id(purpose: :direct_upload)` in `DirectUploadsController`. I think that’ll look like this:

```ruby
def direct_upload_json(blob)
  {
    signed_id: blob.signed_id(purpose: :direct_upload),
    direct_upload: {
        url: blob.service_url_for_direct_upload,
        headers: blob.service_headers_for_direct_upload
    }
  }
end
```

**Review by georgeclaghorn — CHANGES_REQUESTED** [review]

File: activestorage/app/models/active_storage/blob.rb (hunk 1 above)
Comment (santib) [reply activestorage/app/models/active_storage/blob.rb]: Added a commit addressing this comment. Notice that now we won't be sending the following attributes anymore: `id`, `filename`, `byte_size`, `checksum`, `metadata` and `content_type`. Not sure if there was a reason to send them in the past..

Comment (georgeclaghorn) [reply activestorage/app/models/active_storage/blob.rb]: Ah, sorry, I forgot that `as_json` with the `:methods` option includes attributes (and not just the provided methods). We can do this instead:

```ruby
def direct_upload_json(blob)
  blob.as_json(root: false).merge(
    signed_id: blob.signed_id(purpose: :direct_upload),
    direct_upload: {
      # ...
    }
  )
end
```

**Comment by georgeclaghorn** [comment]
I don’t feel comfortable breaking already-issued signed IDs for upgrading apps. I think we need to add a config option, default it on for new apps, and deprecate the old behavior.

**Comment by santib** [comment]
Sounds fair to me @georgeclaghorn. I just updated the PR making the change be configurable based on how I found it being done in other parts of Rails. I still have some concerns:
- This approach still has a non-zero downtime during migration. To have no downtime during the migration we'll need to have the `find_or_build_blob` method find by any of the two purposes for some time. Not sure it's worth it though..
- For people that monkey patched the ActiveStorage controllers (e.g. for security purposes) they MUST update the controller when upgrading.
- For people that used `ActiveStorage::Blob.find_signed(..., purpose: :blob_id)` or similar for direct uploads flows they MUST update their code to use the new purpose.

Any thoughts on the above?

On the other hand, the good thing is that if this is included for Rails 7 people by default will probably be extra cautious (for being a major version upgrade)

**Comment by georgeclaghorn** [comment]
> This approach still has a non-zero downtime during migration. To have no downtime during the migration we'll need to have the find_or_build_blob method find by any of the two purposes for some time. Not sure it's worth it though..

Yeah, this is what I had in mind (and I do think it needs to be seamless). Generate signed IDs with the new purpose unconditionally. The flag controls whether we accept the old purpose as a fallback. Warn in an `after_initialize` hook if configured to accept the old purpose.

File: activestorage/lib/active_storage/engine.rb [hunk 2]
```diff
@@ -157,5 +157,20 @@ class Engine < Rails::Engine # :nodoc:
         ActiveStorage::FixtureSet.file_fixture_path = ActiveSupport::TestCase.file_fixture_path
       end
     end
+
+    initializer "active_storage.old_signed_id_fallback" do
+      config.after_initialize do |app|
+        ActiveStorage.old_signed_id_fallback = app.config.active_storage.old_signed_id_fallback || true
```
Comment (georgeclaghorn) [inline activestorage/lib/active_storage/engine.rb:163]: This will always be `true`, right? Think we can set a default value for `config.active_storage.old_signed_id_fallback` up top and do away with the `|| true`.

**Comment by santib** [comment]
Hi @gmcgibbon I don't want to bug you, but I don't know who to ping now that George left the team. Are you able to help moving this security fix forward or otherwise can you point me to the right person? Thanks 🙌

File: activestorage/lib/active_storage.rb [hunk 3]
```diff
@@ -67,6 +67,7 @@ module ActiveStorage
 
   mattr_accessor :replace_on_assign_to_many, default: false
   mattr_accessor :track_variants, default: false
+  mattr_accessor :old_signed_id_fallback, default: true
```
Comment (gmcgibbon) [inline activestorage/lib/active_storage.rb:70]: Configuration booleans that opt into new behaviour is typically defaulted to false, and explicitly enabled with true.  This makes the logic switches much more readable and idiomatic. Maybe `mattr_accessor :direct_upload_signing`?

File: railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_7_0.rb.tt [hunk 4]
```diff
@@ -37,3 +37,10 @@
 
 # Set both the `:open_timeout` and `:read_timeout` values for `:smtp` delivery method.
 # Rails.application.config.action_mailer.smtp_timeout = 5
+
+# Uses a specific purpose signed id for direct uploads instead of using the same signed id
+# used in other places. Changing this default means that ActiveStorage won't continue to attach
+# blobs by their old signed id `purpose: :blob_id`. It is recommended to first deploy Rails 7
+# with the previous defaults and then uncomment the flag to be `false`.
```
Comment (gmcgibbon) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_7_0.rb.tt:44]:
```
# Sign direct uploads with a unique purpose.
#
# Direct uploads should now signed with a :direct_upload purpose. Previously, they were signed with :blob_id. 
# If you use direct uploads, your app will need to temporarily be able to find both signed types.
# Make sure your entire app is migrated and stable on 7.0 before using this setting.
#
# Rails.application.config.active_storage.direct_upload_purpose_signing = false
```

Somewhat similar to https://github.com/rails/rails/pull/34692/files#diff-a10710b71e54cf750d72df8516aa9ca7a01a4239bdc9b36575208f116cae406a. I think something like this would sound better.

File: activestorage/lib/active_storage/engine.rb [hunk 5]
```diff
@@ -157,5 +158,17 @@ class Engine < Rails::Engine # :nodoc:
         ActiveStorage::FixtureSet.file_fixture_path = ActiveSupport::TestCase.file_fixture_path
       end
     end
+
+    initializer "active_storage.old_signed_id_fallback" do
+      config.after_initialize do |app|
+        if ActiveStorage.old_signed_id_fallback
+          ActiveSupport::Deprecation.warn(<<-MSG.squish)
```
Comment (gmcgibbon) [inline activestorage/lib/active_storage/engine.rb:165]: Typically I think we show deprecation warnings on the behaviour we are switching on. That would be in `lib/active_storage/attached/changes/create_one.rb `. If we find a record with the old signing purpose, we should log a deprecation warning.

File: activestorage/lib/active_storage.rb (hunk 3 above)
Comment (santib) [reply activestorage/lib/active_storage.rb:70]: Sounds fair, although I'm not sure about the name `direct_upload_purpose_signing` since we are always going to use the new purpose for signing, the flag is there only to stop allowing the old purpose (blob_id) as a fallback. What do you think of `ignore_old_purpose_for_direct_uploads` ? By default it would be `false` so we accept both: `blob_id` and `direct_upload` as the purpose.

File: activestorage/lib/active_storage/engine.rb (hunk 5 above)
Comment (brunvez) [reply activestorage/lib/active_storage/engine.rb:165]: Hey thanks for the review! I didn't quite follow this one, are you suggesting that we fire a warning on every blob we find with the old signature or just not fire at all?

Comment (gmcgibbon) [reply activestorage/lib/active_storage/engine.rb:165]: I mean instead of logging a deprecation here we should do it in `lib/active_storage/attached/changes/create_one.rb` instead. Something like this:

```
if ActiveStorage.old_signed_id_fallback
  if (blob = ActiveStorage::Blob.find_signed(attachable, record: record, purpose: :blob_id)
    # warn here
    blob
  else
    ActiveStorage::Blob.find_signed!(attachable, record: record, purpose: :direct_upload)
  end
else
  ActiveStorage::Blob.find_signed!(attachable, record: record, purpose: :direct_upload)
end
```

File: activestorage/lib/active_storage.rb (hunk 3 above)
Comment (gmcgibbon) [reply activestorage/lib/active_storage.rb:70]: Hm, yeah my suggested name isn't quite right.  How about `ignore_blob_id_purpose_for_direct_uploads ` or maybe `find_direct_uploads_by_direct_upload_purpose`? Either of the these are fine with me.

File: activestorage/lib/active_storage/engine.rb (hunk 5 above)
Comment (santib) [reply activestorage/lib/active_storage/engine.rb]: Just in case, when addressing this comment we decided to always find with `purpose: :direct_upload` first and if not found then for `purpose: :blob_id` to prevent an extra query since in normal cases all new signatures will be using `purpose: :direct_upload`

File: activestorage/lib/active_storage/attached/changes/create_one.rb [hunk 6]
```diff
@@ -69,7 +69,21 @@ def find_or_build_blob
...
           )
         when String
-          ActiveStorage::Blob.find_signed!(attachable, record: record)
+          if ActiveStorage.ignore_blob_id_purpose_for_direct_uploads
+            ActiveStorage::Blob.find_signed!(attachable, record: record, purpose: :direct_upload)
+          else
+            if (blob = ActiveStorage::Blob.find_signed(attachable, record: record, purpose: :direct_upload))
+              blob
+            else
+              ActiveStorage::Blob.find_signed!(attachable, record: record, purpose: :blob_id).tap do |blob|
```
Comment (gmcgibbon) [inline activestorage/lib/active_storage/attached/changes/create_one.rb:78]:
```suggestion
              ActiveStorage::Blob.find_signed!(attachable, record: record, purpose: :blob_id).tap
```

File: activestorage/test/models/attached/one_test.rb [hunk 7]
```diff
@@ -629,4 +632,14 @@ def avatar
...
   end
+
+  test "using the old signed id for direct uploads attaches the blob" do
+    @was_ignore_blob_id_purpose_for_direct_uploads = ActiveStorage.ignore_blob_id_purpose_for_direct_uploads
+    ActiveStorage.ignore_blob_id_purpose_for_direct_uploads = false
+
+    @user.avatar.attach create_blob(filename: "funky.jpg").signed_id(purpose: :blob_id)
+    assert_equal "funky.jpg", @user.avatar.filename.to_s
+
+    ActiveStorage.ignore_blob_id_purpose_for_direct_uploads = @was_ignore_blob_id_purpose_for_direct_uploads
```
Comment (gmcgibbon) [inline activestorage/test/models/attached/one_test.rb:643]: Typically the global state reset is ensured with an `ensure`. You can cut down on repeated code by making a helper like this: https://github.com/rails/rails/blob/6ff8725d87dc6f73d0c9401ed9470e7a6dde1409/activestorage/test/test_helper.rb#L103-L110

File: activestorage/lib/active_storage/attached/changes/create_one.rb [hunk 8]
```diff
@@ -69,7 +69,21 @@ def find_or_build_blob
             ).symbolize_keys
           )
         when String
-          ActiveStorage::Blob.find_signed!(attachable, record: record)
```
Comment (gmcgibbon) [inline activestorage/lib/active_storage/attached/changes/create_one.rb:72]: I'm hesitant to believe that this code path will only ever be used for direct uploads. Users could try attaching by signed ID via database seeding for example. There may also be legitimate use-cases for using one blob for multiple attachment relations.

How can we conditionalize this further to only check when the attachable is meant to be a direct upload? The fact that we had to turn on the new feature for all of our one + many attachment tests is an indicator that we're breaking contract on other code paths.

File: railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_7_0.rb.tt [hunk 9]
```diff
@@ -37,3 +37,11 @@
 
 # Set both the `:open_timeout` and `:read_timeout` values for `:smtp` delivery method.
 # Rails.application.config.action_mailer.smtp_timeout = 5
+
+# Sign direct uploads with a specific purpose.
+#
+# Direct uploads should now be signed with a `:direct_upload` purpose. Previously, they were signed with `:blob_id`.
+# If you use direct uploads, your app will need to temporarily be able to find both signed types.
+# Make sure your entire app is migrated and stable on 7.0 before using this setting.
```
Comment (zzak) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_7_0.rb.tt:45]: Should there be a changelog for this?

I'm also curious about:

> If you use direct uploads, your app will need to temporarily be able to find both signed types.
> Make sure your entire app is migrated and stable on 7.0 before using this setting.

If we're going to go down this route of breaking uploads, we should have guidance on how to migrate. IMO

File: activestorage/lib/active_storage/attached/changes/create_one.rb (hunk 8 above)
Comment (brunvez) [reply activestorage/lib/active_storage/attached/changes/create_one.rb:72]: Hey @gmcgibbon, we gave some thought to your comments and we think that breaking the API is inevitable if we are to fix the security vulnerability but we should also contemplate older apps running on production right now. To consider these older apps we agree this should be a configuration to allow them to keep their current behavior and since, as you mentioned, the path is more widely used we thought about renaming the purpose from `:direct_upload` to `:private_id`.

The configuration will be set to use `:private_id` by default on new Rails 7+ apps prioritizing security and `:blob_id` on older ones. To ensure a smooth migration we also included a `:private_id_with_fallback` value which will sign direct uploads with `:private_id` but accept both `:blob_id` and `:private_id` until they change it to the more secure mode `:private_id`. We've included these changes along with an explanation in the CHANGELOG and migration guides as per @zzak's comments :)

File: activestorage/test/controllers/direct_uploads_controller_test.rb [hunk 10]
```diff
@@ -113,7 +113,7 @@ class ActiveStorage::AzureStorageDirectUploadsControllerTest < ActionDispatch::I
         filename: "hello.txt", byte_size: 6, checksum: checksum, content_type: "text/plain", metadata: metadata } }
 
       @response.parsed_body.tap do |details|
-        assert_equal ActiveStorage::Blob.find(details["id"]), ActiveStorage::Blob.find_signed!(details["signed_id"])
+        assert_equal ActiveStorage::Blob.find(details["id"]), ActiveStorage::Blob.find_signed!(details["signed_id"], purpose: :direct_upload)
```
Comment (santib) [inline activestorage/test/controllers/direct_uploads_controller_test.rb:116]: I think we need to revert this change

**Comment by zzak** [comment]
If I'm understanding this correctly, the default behavior is still to sign by `:blob_id` but we provide a configuration flag for users to migrate their applications once they're running on 7.0.

Are we setting the flag to default to `:private_id` for newly generated apps?

**Comment by santib** [comment]
> If I'm understanding this correctly, the default behavior is still to sign by `:blob_id` but we provide a configuration flag for users to migrate their applications once they're running on 7.0.

That's correct. 

> Are we setting the flag to default to `:private_id` for newly generated apps?

Newly generated apps (once Rails 7 is out) will load 7.0 defaults and be affected by https://github.com/rails/rails/pull/41433/files#diff-9f26e965e7e2a6842bd6ea755924a621038bc9dd8d63cfa6b34092aa32ede02cR221

**Comment by zzak** [comment]
@gmcgibbon Could you give this another look too plz? :bow:

**Comment by pixeltrix** [comment]
Given that it's a major version change a lot of apps will want to run 6.1 and 7.0 in production at the same time - is that scenario covered by the `:private_id_with_falback` option or will they need to keep the setting at `:blob_id` for the duration of that phase. If the latter we'd want to point that out in the upgrading guide/changelog.

Can we also add some configuration tests similar to other new framework default tests in `railties/test/application/configuration_test.rb` - thanks.

File: activestorage/test/models/attached/many_test.rb [hunk 11]
```diff
@@ -667,11 +667,50 @@ def highlights
...
     ensure
       ActiveStorage.replace_on_assign_to_many = previous
     end
+
+    def blob_attachment_mode(mode)
+      ActiveStorage.blob_attachment_mode, previous = mode, ActiveStorage.blob_attachment_mode
+      yield
+    ensure
+      ActiveStorage.blob_attachment_mode = previous
+    end
```
Comment (pixeltrix) [inline activestorage/test/models/attached/many_test.rb:715]: Can we rename this to `with_blob_attachment_mode` and extract the two instances of it to `test_helper.rb` in a similar manner to `with_service`, thanks.

**Comment by santib** [comment]
@pixeltrix we just pushed the requested changes. Regarding:

> Given that it's a major version change a lot of apps will want to run 6.1 and 7.0 in production at the same time - is that scenario covered by the `:private_id_with_falback` option or will they need to keep the setting at `:blob_id` for the duration of that phase. If the latter we'd want to point that out in the upgrading guide/changelog.

Apps could be running in dual boot way with `:private_id_with_falback` without issues because the blob ids signed with any of the two purposes (`:blob_id` or `:private_id`) will be accepted when attaching them.
